server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,310 @@
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+import org.elasticsearch.health.SimpleHealthIndicatorDetails;
+import org.elasticsearch.health.UserAction;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.health.ServerHealthComponents.CLUSTER_COORDINATION;
+
+/**
+ * This indicator reports the health of master stability.
+ * If we have had a master within the last 30 seconds, and that master has not changed more than 3 times in the last 30 minutes, then
+ * this will report GREEN.
+ * If we have had a master within the last 30 seconds, but that master has changed more than 3 times in the last 30 minutes (and that is
+ * confirmed by checking with the last-known master), then this will report YELLOW.
+ * If we have not had a master within the last 30 seconds, then this will will report RED with one exception. That exception is when:
+ * (1) no node is elected master, (2) this node is not master eligible, (3) some node is master eligible, (4) we ask a master-eligible node
+ * to run this indicator, and (5) it comes back with a result that is not RED.
+ * Since this indicator needs to be able to run when there is no master at all, it does not depend on the dedicated health node (which
+ * requires the existence of a master).
+ */
+public class StableMasterHealthIndicatorService implements HealthIndicatorService, ClusterStateListener {
+
+    public static final String NAME = "stable_master";
+
+    private final ClusterService clusterService;
+    private final MasterHistoryService masterHistoryService;
+    private static final Logger logger = LogManager.getLogger(StableMasterHealthIndicatorService.class);
+    /**
+     * This is the amount of time we use to make the initial decision -- have we seen a master node in the very recent past?
+     */
+    private static final TimeValue VERY_RECENT_PAST = new TimeValue(30, TimeUnit.SECONDS);
+
+    /**
+     * This is the number of times that it is OK for the master history to show a transition from a non-null master to a null master before
+     * it starts impacting the health status.
+     */
+    private static final int ACCEPTABLE_NULL_TRANSITIONS = 3;
+    /**
+     * This is the number of times that it is OK for the master history to show a transition one non-null master to a different non-null
+     * master before it starts impacting the health status.
+     */
+    private static final int ACCEPTABLE_IDENTITY_CHANGES = 3;
+    /**
+     * The severity to use for a low-severity impact in this indicator
+     */
+    private static final int LOW_IMPACT_SEVERITY = 3;
```

Comment:
I think we should configure things like this with (static node-level) settings. We have access to such settings via `ClusterService#getSettings`. We think these are good defaults, but if we discover they're too sensitive in some installations then it would be useful to let folks change them.

Reply:
Done. I don't know if we have any naming guidelines for settings. I've tried to match patterns I've seen in the code but there's a lot of variety.